BUG: fix denormal linspace test for longdouble

matthew-brett opened the pull request:
Linspace test was testing for case where simply dividing the range by
the number of samples generates 0, as the case for very small denormals.
The implementation generating the denormal only worked to exercise that
code for float64.  Make a small denormal that does exercise that case,
as least for longdouble, as well as float64.

Closes gh-8565.

matthew-brett committed:
BUG: fix denormal linspace test for longdouble

Linspace test was testing for case where simply dividing the range by
the number of samples generates 0, as the case for very small denormals.
The implementation generating the denormal only worked to exercise that
code for float64.  Make a small denormal that does exercise that case,
as least for longdouble, as well as float64.

Closes gh-8565.

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -301,9 +301,9 @@ def __mul__(self, other):
-            stop = finfo(dt).tiny * finfo(dt).resolution
-            assert_(any(linspace(0, stop, 10, endpoint=False, dtype=dt)))
+        for ftype in sctypes['float']:
+            stop = nextafter(ftype(0), ftype(1)) * 5  # A denormal number
```
If this was 0 before, then an assertion (or perhaps better, an exception to make the test error) that it is not zero now would probably be a good thing.

matthew-brett replied:
I don't think it can be zero now, unless there's a bug in `nextafter`.

eric-wieser replied:
Testing for a bug in `nextafter` sounds like something that should be tested somewhere, and if not elsewhere, here is as good a place as any

matthew-brett committed:
TST: add tests of nextafter around 0

Check that nextafter(0) gives number in range ]0 ... tiny[, and number
divided by 2 is 0 (denormal).

matthew-brett commented:
I added some tests for nextafter with the other nextafter tests.

eric-wieser commented on numpy/core/tests/test_umath.py:
```diff
@@ -2126,6 +2126,12 @@ def test_nextafterf():
+def test_nextafter_0():
+    for t, direction in itertools.product(np.sctypes['float'], (1, -1)):
+        tiny = np.finfo(t).tiny
+        assert_(0. < direction * np.nextafter(t(0), t(direction)) < tiny)
```
is `tiny` guaranteed not to be equal to `nextafter(0, 1)`?

matthew-brett replied:
Yes, for all floats we support, because they all support denormal numbers.  `tiny` is the smallest not-denormal number the type supports.

eric-wieser replied:
This test fails on window subsystem for linux

eric-wieser replied:
Because `1*np.nextafter(t(0), t(direction))` is `0`!

eric-wieser replied:
#8726

eric-wieser commented on numpy/core/tests/test_umath.py:
```diff
@@ -2126,6 +2126,12 @@ def test_nextafterf():
+    for t, direction in itertools.product(np.sctypes['float'], (1, -1)):
+        tiny = np.finfo(t).tiny
+        assert_(0. < direction * np.nextafter(t(0), t(direction)) < tiny)
+        assert_equal(np.nextafter(t(0), t(direction)) / t(2), direction * 0.0)
```
And is this guaranteed to round down?

matthew-brett replied:
Hum - good question.  I'll divide by 2.1 to be sure.

eric-wieser reviewed (approved):
Some queries about why this should pass, but assuming this does pass, looks good to me

eric-wieser commented:
Thanks @matthew-brett
